File: tools/tapscore-watch/WatchApp/RemoteModel.swift
import Foundation
import Observation
import WatchConnectivity

// "Remote" mode: the watch controls the match running on the phone. It sends score/undo/sync
// commands and mirrors the state the phone broadcasts. The phone stays the source of truth.
@Observable
final class RemoteModel: NSObject, WCSessionDelegate {
    var reachable = false          // phone app is in the foreground and reachable
    var active = false             // phone has a match on-screen to control
    var over = false
    var winner = -1                // 0/1, or -1 tie
    var you = "0"
    var opp = "0"
    var pill = ""
    var server = 0
    var names: [String] = ["You", "Opp"]

    override init() {
        super.init()
        if WCSession.isSupported() {
            let s = WCSession.default
            s.delegate = self
            s.activate()
        }
    }

    // MARK: commands (watch → phone)
    private func send(_ dict: [String: Any]) {
        guard WCSession.isSupported() else { return }
        let s = WCSession.default
        if s.isReachable { s.sendMessage(dict, replyHandler: nil, errorHandler: nil) }
    }
    private func cmd(_ o: [String: Any]) -> [String: Any] {
        let json = (try? JSONSerialization.data(withJSONObject: o)).flatMap { String(data: $0, encoding: .utf8) } ?? "{}"
        return ["json": json]
    }
    func score(_ side: Int) { Haptics.point(); send(cmd(["t": "score", "side": side])) }
    func undo() { Haptics.undo(); send(cmd(["t": "undo"])) }
    func requestSync() { send(cmd(["t": "sync"])) }

    // MARK: state (phone → watch)
    private func apply(_ json: String) {
        guard let d = json.data(using: .utf8),
              let o = try? JSONSerialization.jsonObject(with: d) as? [String: Any] else { return }
        DispatchQueue.main.async {
            self.active = o["active"] as? Bool ?? false
            self.over = o["over"] as? Bool ?? false
            self.winner = o["winner"] as? Int ?? -1
            self.you = o["you"] as? String ?? "0"
            self.opp = o["opp"] as? String ?? "0"
            self.pill = o["pill"] as? String ?? ""
            self.server = o["server"] as? Int ?? 0
            if let n = o["names"] as? [String], n.count == 2 { self.names = n }
        }
    }

    // MARK: WCSessionDelegate
    func session(_ s: WCSession, activationDidCompleteWith state: WCSessionActivationState, error: Error?) {
        DispatchQueue.main.async { self.reachable = (state == .activated && s.isReachable) }
        if state == .activated { requestSync() }
    }
    func sessionReachabilityDidChange(_ s: WCSession) {
        DispatchQueue.main.async { self.reachable = s.isReachable }
        if s.isReachable { requestSync() }
    }
    func session(_ s: WCSession, didReceiveMessage message: [String: Any]) {
        if let j = message["json"] as? String { apply(j) }
    }
    func session(_ s: WCSession, didReceiveApplicationContext ctx: [String: Any]) {
        if let j = ctx["json"] as? String { apply(j) }
    }
}

File: tools/tapscore-wear/WearApp/RemoteModel.kt
package com.ivogomes.tapscore.wear

import android.content.Context
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.google.android.gms.wearable.MessageClient
import com.google.android.gms.wearable.MessageEvent
import com.google.android.gms.wearable.Wearable
import org.json.JSONObject

// "Remote" mode: the watch controls the match on the phone. Sends score/undo/sync over the Wearable
// Data Layer and mirrors the state the phone broadcasts. The phone stays the source of truth.
class RemoteModel(context: Context) : MessageClient.OnMessageReceivedListener {
    private val app = context.applicationContext
    private val msg = Wearable.getMessageClient(app)
    private val nodes = Wearable.getNodeClient(app)
    private val haptics = Haptics(context)
    private val path = "/tapscore"

    var reachable by mutableStateOf(false); private set
    var active by mutableStateOf(false); private set
    var over by mutableStateOf(false); private set
    var winner by mutableStateOf(-1); private set
    var you by mutableStateOf("0"); private set
    var opp by mutableStateOf("0"); private set
    var pill by mutableStateOf(""); private set
    var server by mutableStateOf(0); private set
    var names by mutableStateOf(listOf("You", "Opp")); private set

    fun start() { msg.addListener(this); requestSync() }
    fun stop() { try { msg.removeListener(this) } catch (e: Exception) {} }

    private fun sendCmd(json: String) {
        nodes.connectedNodes
            .addOnSuccessListener { list -> reachable = list.isNotEmpty(); for (n in list) msg.sendMessage(n.id, path, json.toByteArray()) }
            .addOnFailureListener { reachable = false }
    }
    fun score(side: Int) { haptics.point(); sendCmd("""{"t":"score","side":$side}""") }
    fun undo() { haptics.undo(); sendCmd("""{"t":"undo"}""") }
    fun requestSync() { sendCmd("""{"t":"sync"}""") }

    override fun onMessageReceived(e: MessageEvent) {
        if (e.path != path) return
        try {
            val o = JSONObject(String(e.data))
            reachable = true
            active = o.optBoolean("active"); over = o.optBoolean("over"); winner = o.optInt("winner", -1)
            you = o.optString("you", "0"); opp = o.optString("opp", "0"); pill = o.optString("pill", "")
            server = o.optInt("server", 0)
            o.optJSONArray("names")?.let { if (it.length() == 2) names = listOf(it.getString(0), it.getString(1)) }
        } catch (ex: Exception) {}
    }
}
